File: app/src/main/java/com/suhang/networkmvp/function/FlowableWrap.java
package com.suhang.networkmvp.function;

import com.suhang.networkmvp.domain.ErrorBean;
import com.suhang.networkmvp.event.result.BaseResult;
import com.suhang.networkmvp.event.ErrorCode;
import com.suhang.networkmvp.utils.LogUtil;

import io.reactivex.Flowable;
import io.reactivex.disposables.CompositeDisposable;

/**
 * Created by 苏杭 on 2017/4/28 11:02.
 */

public class FlowableWrap<T extends BaseResult> {
    private Flowable<T> mFlowable;
    private CompositeDisposable mDisposable;

    public FlowableWrap(Flowable<T> flowable, CompositeDisposable disposable) {
        mFlowable = flowable;
        mDisposable = disposable;
    }


    public FlowableWrap<T> subscribe(Next<T> consumer) {
        mDisposable.add(mFlowable.subscribe(consumer::onNext, throwable -> LogUtil.i("啊啊啊" + new ErrorBean(ErrorCode.ERROR_CODE_SUBSCRIBE_INNER, ErrorCode.ERROR_DESC_SUBSCRIBE_INNER))));
        return this;
    }

    public FlowableWrap<T> subscribe(Next<T> consumer, Error error) {
        mDisposable.add(mFlowable.subscribe(consumer::onNext, error::onError));
        return this;
    }

    public interface Next<T> {
        void onNext(T t);
    }

    public interface Error {
        void onError(Throwable throwable);
    }

}

File: app/src/main/java/com/suhang/networkmvp/function/FlowableWrap.kt
package com.suhang.networkmvp.function

import com.suhang.networkmvp.constants.ErrorCode
import com.suhang.networkmvp.constants.errorMessage
import com.suhang.networkmvp.domain.ErrorBean
import com.suhang.networkmvp.utils.LogUtil

import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.warn

/**
 * Created by 苏杭 on 2017/4/28 11:02.
 */

class FlowableWrap<T>(private val mFlowable: Flowable<T>, private val mDisposable: CompositeDisposable):AnkoLogger{


    fun subscribe(consumer: Next<T>): FlowableWrap<T> {
        mDisposable.add(mFlowable.subscribe({ consumer.onNext(it) }, {
            warn(ErrorBean(ErrorCode.ERROR_CODE_SUBSCRIBE_INNER, ErrorCode.ERROR_DESC_SUBSCRIBE_INNER+"\n",errorMessage(it)))
        }))
        return this
    }

    fun subscribe(consumer: Next<T>, error: Error): FlowableWrap<T> {
        mDisposable.add(mFlowable.subscribe({ consumer.onNext(it) }, { error.onError(it) }))
        return this
    }

    interface Next<T> {
        fun onNext(t: T)
    }

    interface Error {
        fun onError(throwable: Throwable)
    }

}
